cpp/src/decisiontree/batched-levelalgo/kernels/builder_kernels.cuh

```diff
@@ -111,6 +122,266 @@ HDI IdxT lower_bound(DataT* array, IdxT len, DataT element)
+
+    __syncthreads();
+
+    // do a scan on the mask to get the indices for gathering
+    BlockScanT(temp_storage.scan).ExclusiveSum(mask, col_indices, n_uniques);
+
+  } while (n_uniques < k);
+
+  // write the items[] of only the ones with mask[]=1 to col[offset + col_idx[]]
+  IdxT col_offset = k * blockIdx.x;
+  for (int i = 0; i < MAX_SAMPLES_PER_THREAD; ++i) {
+    if (mask[i] and col_indices[i] < k) { colids[col_offset + col_indices[i]] = items[i]; }
+  }
+}
+
+// algo L of the reservoir sampling algorithm
+/**
+ * @brief Generates 'k' unique samples of features from 'n' feature sample-space using the algo-L
+ * algorithm of reservoir sampling. wiki :
+ * https://en.wikipedia.org/wiki/Reservoir_sampling#An_optimal_algorithm
+ */
+template <typename IdxT>
+__global__ void algo_L_sample_kernel(int* colids,
+                                     const NodeWorkItem* work_items,
+                                     size_t work_items_size,
+                                     IdxT treeid,
+                                     uint64_t seed,
+                                     size_t n /* total cols to sample from*/,
+                                     size_t k /* cols to sample */)
+{
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  if (tid >= work_items_size) return;
+  const uint32_t nodeid = work_items[tid].idx;
+  uint64_t subsequence  = (uint64_t(treeid) << 32) | uint64_t(nodeid);
+  raft::random::PCGenerator gen(seed, subsequence, uint64_t(0));
+  raft::random::UniformIntDistParams<IdxT, uint64_t> uniform_int_dist_params;
+  uniform_int_dist_params.start = 0;
+  uniform_int_dist_params.end   = k;
+  uniform_int_dist_params.diff =
+    uint64_t(uniform_int_dist_params.end - uniform_int_dist_params.start);
+  float fp_uniform_val;
+  IdxT int_uniform_val;
+  // fp_uniform_val will have a random value between 0 and 1
+  gen.next(fp_uniform_val);
+  double W = raft::myExp(raft::myLog(fp_uniform_val) / k);
+
+  size_t col(0);
+  // initially fill the reservoir array in increasing order of cols till k
+  while (1) {
+    colids[tid * k + col] = col;
+    if (col == k - 1)
+      break;
+    else
+      ++col;
+  }
+  // randomly sample from a geometric distribution
+  while (col < n) {
+    // fp_uniform_val will have a random value between 0 and 1
+    gen.next(fp_uniform_val);
+    col += static_cast<int>(raft::myLog(fp_uniform_val) / raft::myLog(1 - W)) + 1;
+    if (col < n) {
+      // int_uniform_val will now have a random value between 0...k
+      raft::random::custom_next(gen, &int_uniform_val, uniform_int_dist_params, IdxT(0), IdxT(0));
+      colids[tid * k + int_uniform_val] = col;  // the bad memory coalescing here is hidden
+      // fp_uniform_val will have a random value between 0 and 1
+      gen.next(fp_uniform_val);
+      W *= raft::myExp(raft::myLog(fp_uniform_val) / k);
+    }
+  }
+}
+
+template <typename IdxT>
+__global__ void adaptive_sample_kernel(int* colids,
```

Comment:
We are keeping the adaptive sampling kernel because we plan to repurpose it for weighted sampling in future.